command/agent/job_endpoint.go

```diff
@@ -358,6 +367,210 @@ func (s *HTTPServer) jobActions(resp http.ResponseWriter, req *http.Request, job
 	return out.Actions, nil
 }
 
+func (s *HTTPServer) jobRunAction(resp http.ResponseWriter, req *http.Request, jobID string) (interface{}, error) {
+
+	s.logger.Info("jobRunAction called")
+
+	// Turns out websocket conncetions must always start with a GET, not a POST. D'oh!
+	// var args structs.JobRunActionRequest
+	// if err := decodeBody(req, &args); err != nil {
+	// 	return nil, CodedError(400, err.Error())
+	// }
+
+	// Build the request and parse the ACL token
+	task := req.URL.Query().Get("task")
+	group := req.URL.Query().Get("group")
+	action := req.URL.Query().Get("action")
+	allocID := req.URL.Query().Get("allocID")
+
+	s.logger.Info("jobRunAction called with task: %s, group: %s, action: %s, allocID: %s", task, group, action, allocID)
+
+	args := structs.JobRunActionRequest{
+		JobID:     jobID,
+		TaskGroup: group,
+		Task:      task,
+		Action:    action,
+		AllocID:   allocID,
+	}
+
+	s.parse(resp, req, &args.QueryOptions.Region, &args.QueryOptions)
+
+	s.logger.Info("jobRunAction called with args: %v", args)
+
+	// TODO: Validate: 1. alloc_exec permission, 2. Job ID, Group Name, Task Name, and Action Name are valid
+	// (Can use the job-level action lookup to do this most effectively, probably!)
+	// or: maybe, should group/task/action lookup all happen in the RPC handler?
+
+	// conn, err := s.wsUpgrader.Upgrade(resp, req, nil)
+	// FIXME: this is an open checkOrigin here that allows :4200 to make requests to :4646,
+	// freeing local ember up from not having to proxy.
+	// This is like three workarounds in a trenchcoat and I dno't feel good about it but it unblocks me
+
+	var upgrader = websocket.Upgrader{
+		// Allow all origins
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	s.logger.Info("jobRunAction called with upgrader: %v", upgrader)
+
+	// Then when you upgrade the connection:
+	conn, err := upgrader.Upgrade(resp, req, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to upgrade connection: %v", err)
+	}
+
+	s.logger.Info("jobRunAction thru upgrade")
+
+	if err := s.readActionWsHandshake(conn.ReadJSON, req, &args.QueryOptions); err != nil {
+		conn.WriteMessage(websocket.CloseMessage,
+			websocket.FormatCloseMessage(toWsCode(400), err.Error()))
+		return nil, err
+	}
+
+	s.logger.Info("jobRunAction thru handshake")
+
+	return s.runActionStreamImpl(conn, &args)
```

Comment:
I'd strongly consider at this point whether we can call `s.execStreamImpl` here without reimplementing a copy of it.